### PR TITLE
Feat/issue 68 hybrid search

### LINKED ISSUE #68: Add hybrid search
Encorporate Qdrant hybrid search with RRF

### PR DESCRIPTION
**Pull request by loganmoon**
closes #68

### CODE REVIEW

**Comment by loganmoon** [comment]
## Code Review Summary

This PR implements hybrid search with BM25 + RRF fusion. The implementation is well-structured with good abstractions, but has **critical test coverage gaps** and several code issues that should be addressed before merging.

### 🔴 Critical Issues

**1. Test Coverage Gaps**

The PR adds ~1,328 lines but lacks tests for core functionality:

- **RRF Fusion Logic** - No tests validating the RRF algorithm, prefetch multiplier, or result ranking
- **BM25 Statistics** - No tests for avgdl calculation accuracy, incremental updates, or deletion scenarios  
- **Hybrid Search E2E** - No end-to-end test validating complete query flow
- **Edge Cases** - Missing tests for avgdl=0, zero entities, empty inputs, sparse embedding failures

**Recommendation:** Add integration tests for:
- RRF fusion in `crates/storage/tests/qdrant_integration_test.rs`
- BM25 statistics in `crates/storage/tests/postgres_integration_test.rs`
- E2E hybrid search in `tests/tests/`

See inline comments for specific code issues.

### ✅ Positive Observations

- Well-tested tokenizer (8 comprehensive tests)
- Clean separation via traits
- Incremental avgdl updates
- Sensible configuration defaults

### Recommendation

**DO NOT MERGE** until test coverage gaps and inline code issues are addressed.

**Review by loganmoon — COMMENTED** [review]
**Issue #1: Integer Overflow Risk in Token Count Accumulation** 🔴

**File:** `crates/storage/src/postgres/client.rs`
**Function:** `update_bm25_statistics_incremental()`

**Problem:**
```rust
let new_total_tokens: i64 = new_token_counts.iter().sum::<usize>() as i64;
```

Direct cast from `usize` to `i64` can silently overflow if the sum exceeds `i64::MAX`. Large repositories could hit this limit, causing silent data corruption and incorrect avgdl calculations.

**Fix:**
```rust
let new_total_tokens: i64 = new_token_counts
    .iter()
    .try_fold(0i64, |acc, &count| {
        acc.checked_add(count as i64)
            .ok_or_else(|| Error::msg("Token count overflow"))
    })?;
```

**Priority:** High (data integrity issue)

**Review by loganmoon — COMMENTED** [review]
**Issue #2: Unnecessary String Cloning in Sparse Embedding Pipeline** 🔴

**File:** `crates/indexer/src/entity_processor.rs`
**Function:** `process_entity_chunk()`

**Problem:**
```rust
let all_entity_contents: Vec<String> = entity_contents_and_hashes
    .iter()
    .map(|(content, _)| content.clone())  // Unnecessary clone
    .collect();

let all_sparse_embeddings = sparse_manager
    .embed_sparse(all_entity_contents)
    .await
```

This clones every entity's embedding content (potentially large strings) when the sparse provider only needs read-only access.

**Impact:**
- 2x memory usage for entity content during sparse embedding
- O(n) allocations where n = batch size (up to hundreds of entities)

**Fix:**
Change `SparseEmbeddingProvider::embed_sparse()` signature to accept borrowed strings:

```rust
// In sparse_provider.rs trait
async fn embed_sparse(&self, texts: Vec<&str>) 
    -> Result<Vec<Option<Vec<(u32, f32)>>>>;

// In entity_processor.rs
let all_entity_contents: Vec<&str> = entity_contents_and_hashes
    .iter()
    .map(|(content, _)| content.as_str())
    .collect();
```

**Priority:** High (significant memory waste during indexing)

**Review by loganmoon — COMMENTED** [review]
**Additional Observation: Token Count Type Inconsistency** ℹ️

**Files:** Multiple files across crates

**Issue:**
Token counts flow through three different integer types:
1. `calculate_token_counts()` returns `Vec<usize>`
2. Database stores as `INTEGER` (i32) 
3. Statistics calculations use `i64`

**Potential Problem:**
- Truncation possible: `usize` (64-bit) → `i32` (32-bit, max ~2.1B tokens)
- Unlikely but possible for extremely large files (100MB+ source files)

**Suggested Fix (Low Priority):**
Use `u32` consistently and document the limitation, or use `BIGINT` in database schema if concerned about future-proofing:

```rust
pub fn calculate_token_counts(entities: &[CodeEntity]) -> Result<Vec<u32>> {
    // ...
    token_counts.push(tokens.len() as u32);  // Explicit truncation
}
```

**Priority:** Low (document limitation or validate entity sizes during indexing)

File: crates/storage/src/postgres/client.rs [hunk 1]
```diff
@@ -283,6 +284,141 @@ impl PostgresClient {
...
+
+        Ok(super::BM25Statistics {
+            avgdl: row.0.unwrap_or(50.0),
+            total_tokens: row.1.unwrap_or(0),
+            entity_count: row.2.unwrap_or(0),
+        })
+    }
+
+    /// Update BM25 statistics incrementally after adding new entities
+    pub async fn update_bm25_statistics_incremental(
```
Comment (loganmoon) [inline crates/storage/src/postgres/client.rs]: It seems like these update functions should be handled by outbox_processor inside its write transaction.

File: crates/server/src/mcp_server.rs [hunk 2]
```diff
@@ -185,14 +186,81 @@ impl CodeSearchMcpServer {
                 )
             })?;
 
-        let query_embedding = embeddings.into_iter().next().flatten().ok_or_else(|| {
+        let dense_query_embedding = embeddings.into_iter().next().flatten().ok_or_else(|| {
```
Comment (loganmoon) [inline crates/server/src/mcp_server.rs]: This is unnecessarily complicated since there must be exactly one embedding vector.
